spark/src/main/java/org/apache/iceberg/spark/actions/BaseRewriteDataFilesSparkAction.java

```diff
@@ -0,0 +1,321 @@
+            } catch (Exception e) {
+              if (!partialProgressEnabled) {
+                LOG.error("Failure during rewrite commit process, partial progress not enabled. Rethrowing", e);
+                throw e;
+              } else {
+                LOG.error("Failure during rewrite commit process, partial progress enabled. Ignoring", e);
+              }
+            }
+            completedCommits.addAll(batch);
+          }
+        }
+        Set<String> finalBatch = Sets.newHashSetWithExpectedSize(groupsPerCommit);
+        while (completedRewrite.size() != 0) {
+          finalBatch.add(completedRewrite.poll());
+        }
+        if (finalBatch.size() != 0) {
+          // Final Commit
+          commitFileGroups(finalBatch);
+          completedCommits.addAll(finalBatch);
+        }
+      } catch (Exception e) {
+        if (!partialProgressEnabled) {
+          LOG.error("Cannot commit rewrite and partial progress not enabled, removing all completed work", e);
+          completedCommits.forEach(this::abortFileGroup);
+        }
+      }
+    });
+
+    // Start rewrite tasks
+    Tasks.Builder<Pair<FileGroupInfo, List<FileScanTask>>> rewriteTaskBuilder = Tasks.foreach(jobStream.iterator())
+        .suppressFailureWhenFinished()
+        .executeWith(rewriteService)
+        .noRetry()
+        .onFailure((info, exception) -> {
+          LOG.error("Failure during rewrite process for group {}", info.first(), exception);
+          abortFileGroup(info.first().setId);
+        });
+
+    if (partialProgressEnabled) {
+      rewriteTaskBuilder = rewriteTaskBuilder.suppressFailureWhenFinished();
+    } else {
+      rewriteTaskBuilder = rewriteTaskBuilder.stopOnFailure();
+    }
+
+    rewriteTaskBuilder
+        .run(i -> {
+          String setId = i.first().setId();
+          Set<DataFile> addedFiles = strategy.rewriteFiles(setId, i.second());
+          completedRewrite.offer(setId);
+          results.put(setId, Pair.of(i.first(), new FileGroupRewriteResult(addedFiles.size(), i.second().size())));
+        });
+    stillRewriting.set(false);
+    commiterService.shutdown();
+
+    try {
+      commiterService.awaitTermination(10, TimeUnit.MINUTES);
+    } catch (InterruptedException e) {
+      throw new RuntimeException(e);
+    }
+
+    return new Result(results.values().stream().collect(Collectors.toMap(p -> p.first(), p -> p.second())));
+  }
+
+  protected Table table() {
+    return table;
+  }
+
+  /**
+   * Perform a commit operation on the table adding and removing files as required for this set of file groups,
+   * on failure should clean up and rethrow exception
+   * @param completedSetIds fileSets to commit
+   */
+  protected abstract void commitFileGroups(Set<String> completedSetIds);
+
+  /**
+   * Clean up a specified file set by removing any files created for that operation
+   * @param setId fileSet to clean
+   */
+  protected abstract void abortFileGroup(String setId);
```

Comment:
Let's move this directly after the constructor together with other abstract methods.